pulser/simulation/simresults.py

```diff
@@ -60,6 +67,205 @@ def __init__(self, run_output: Sequence[qutip.Qobj], dim: int, size: int,
+        """Probability distribution of the bitstrings"""
+        return self._states
+
+    def get_state(self, t: int) -> qutip.Qobj:
+        """Get the state at time t of the simulation as a diagonal density
+        matrix.
+        This is not the density matrix of the system, but is a convenient way
+        of computing expectation values of observables.
+
+        Args:
+            t (int): index of the state to be returned.
+
+        Returns:
+            qutip.Qobj: States probability distribution as a density matrix.
+        """
+        def _proj_from_bitstring(bitstring: str) -> qutip.Qobj:
+            # In the digital case, |h> = |1> = qutip.basis()
+            # 'all' basis is unacceptable here after projection on bitstrings
+            if self._meas_basis == 'digital':
+                proj = qutip.tensor([qutip.basis(2, int(i)).proj() for i
+                                     in bitstring])
+            # ground-rydberg measurement basis case
+            else:
+                proj = qutip.tensor([qutip.basis(2, 1-int(i)).proj() for i
+                                     in bitstring])
+            return proj
+
+        return sum(v * _proj_from_bitstring(b) for
+                   b, v in self._states[t].items())
+
+    def get_final_state(self) -> qutip.Qobj:
+        """Get the final state of the simulation as a diagonal density matrix.
+        This is not the density matrix of the system, but is a convenient way
+        of computing expectation values of observables.
+
+        Returns:
+            qutip.Qobj: States probability distribution as a density matrix.
+        """
+        return self.get_state(-1)
+
+    def expect(self, obs_list: Sequence[Union[qutip.Qobj, ArrayLike]]
+               ) -> list[Union[float, complex, ArrayLike]]:
+        """Calculates the expectation value of a list of observables.
+
+        Args:
+            obs_list (array-like of qutip.Qobj or array-like of numpy.ndarray):
+                A list of observables whose expectation value will be
+                calculated. If necessary, each member will be transformed into
+                a qutip.Qobj instance.
+        Returns:
+            list: the list of expectation values of each operator.
+        """
+        density_matrices = [self.get_state(t) for
+                            t in range(len(self._states))]
+        if not isinstance(obs_list, (list, np.ndarray)):
+            raise TypeError("`obs_list` must be a list of operators")
+
+        qobj_list = []
+        for obs in obs_list:
+            if not (isinstance(obs, np.ndarray)
+                    or isinstance(obs, qutip.Qobj)):
+                raise TypeError("Incompatible type of observable.")
+            if obs.shape != (2**self._size, 2**self._size):
+                raise ValueError("Incompatible shape of observable.")
+            qobj_list.append(qutip.Qobj(obs))
+
+        return qutip.expect(qobj_list, density_matrices)
```

Comment:
Try casting the output to `ArrayLike`:
```suggestion
        return cast(ArrayLike, qutip.expect(qobj_list, density_matrices))
```